version.js

```diff
@@ -1,92 +1,291 @@
   }
-  // Get the latest release
-  const parsed = await getPreviousReleaseVersion(context);
-  const nextVersion = parsed.inc("minor");
-  const timestamp = await getTimestamp(context);
-  if (
-    context.eventName === "push" &&
-    (context.ref === "refs/heads/master" || context.ref === "refs/heads/main")
-  ) {
-    // If we're building the main branch, don't include the `+{shortHash}` part as Python considers this a "local" version
-    // and will not allow it to be uploaded to PyPI.
-    return `${nextVersion.version}-alpha.${timestamp}`;
+
+  if (eventName === "pull_request") {
+    // pull_request events only
+    const baseRef = context.payload?.pull_request?.base?.ref;
+    const prLabels = context.payload?.pull_request?.labels;
+    localDebug(`PR pushed: ${context.eventName} ${context.ref}`);
+    localDebug(`pull_request.base.ref: ${baseRef}`);
+    localDebug(`pull_request.labels: ${JSON.stringify(prLabels)}`);
+
+    const asVersion = tryParseVersionBranch(baseRef);
+    let nextVersion;
+    if (asVersion !== undefined) {
+      localDebug(`Version branch PR: ${baseRef}`);
+      nextVersion = new SemVer(`${asVersion}.0.0`);
+    } else {
+      const previousRelease = await getLatestReleaseVersion(context.repo);
+      const increment = getIncrementTypeFromLabels(prLabels);
+      nextVersion = previousRelease.inc(increment);
+    }
+    const timestamp = await getCommitTimestamp(context.repo, sha);
+    const shortHash = context.sha.slice(0, 7);
+    return `${nextVersion.version}-alpha.${timestamp}+${shortHash}`;
   }
-  const shortHash = context.sha.slice(0, 7);
-  return `${nextVersion.version}-alpha.${timestamp}+${shortHash}`;
+
+  throw new Error(`Unsupported event: ${eventName}`);
+}
+
+/**
+ * @param {SemVer} baseVersion
+ * @param {string} timestamp
+ * @returns {string}
+ */
+function alphaVersion(baseVersion, timestamp) {
+  // Include the short commit hash for pull-requests and other branches to ensure a unique version per commit.
+  // This is considered a "local" version by Python and is not able to be uploaded to PyPI.
+  return `${baseVersion}-alpha.${timestampToUnix(timestamp)}`;
 }
 
-async function getPreviousReleaseVersion(context) {
+/**
+ * @param {string} baseVersion
+ * @param {string} timestamp
+ * @param {string} sha
+ * @returns {string}
+ */
+function localAlphaVersion(baseVersion, timestamp, sha) {
+  return `${baseVersion}-alpha.${timestampToUnix(timestamp)}+${shortHash(sha)}`;
+}
+
+/**
+ * @param {string} ref
+ * @returns {string}
+ */
+function calculateTagVersion(ref) {
+  // Get the version from the tag
+  const tag = ref.replace("refs/tags/", "");
+  localDebug(`tag: ${tag}`);
+  // Ensure it's a valid semver version
+  const parsed = new SemVer(tag);
+  return parsed.version;
+}
+
+/**
+ * @param {string} branchName
+ * @returns {number | undefined}
+ */
+function tryParseVersionBranch(branchName) {
+  const match = branchName.match(/^v(\d+)$/);
```

Comment:
future-proof by supporting major versions > 9 :)

Reply:
The `\d+` means one or more numbers - so double digits works okay. Updated a test to show this works.

Reply:
I know, I was just smiling about how it _is_ future-proof

Reply:
Ah lol 😄